### DISCUSSION TITLE
Shade state improperly assumed

### DISCUSSION THREAD

**Discussion by gymrat801** [aarons22/homebridge-bond#199] [Ideas]
I absolutely love this plugin for my fans and shades, and in almost every way it excels beyond Somfy's own MyLink system. However, when combined with HomeKit there's a logic problem that creates issues in a home where devices are controlled by multiple systems. 

When your home has a lot of shades, for example, its common to use wall switches/remotes routinely for certain situations, but then HomeKit scenes for others, such as Good Night. The Bond plugin, unfortunately, is storing the assumed shade state based on the last position the plugin sent to the device. It can't take into account state changes that occurred because of remotes, IFTTT, etc. The result is that shades will be open, HomeKit thinks they're closed, you tell Siri to set Good Night expecting the shades to all close, but nothing happens. With other plugins that I've used for my Somfy shades this has not been the case. Even if the shades are closed, the close command still gets sent again and this combats the mismatched physical state and HomeKit assumed state.

Has anyone run into this or know of a workaround short of begging for an update to the plugin so that it works similarly to others? I  absolutely love the way Bond controls Somfy as compared to Somfy's own MyLink device...but this state mismatch issue is a deal breaker and it really bums me out.

Thank you so much!

**Comment by chrismerck** [comment]
Does anyone know where is the logic which says _hey, the shade is reporting as already closed, so let's not bother to send a "Close" intent_ ?

Is this a HomeKit thing? Or does the `homebridge-bond` plugin gate it?

The Bond Bridge is not it. If Bond Bridge receives a `Close` action, it will transmit the corresponding signal, regardless of the "state" that the Bridge thinks it's in. [EDIT: Some may be familiar with the "trust_state" behavior for Ceiling Fan lights, where the Bridge *does* ignore TurnLightOn/Off signals when the state is already where desired but only for these CF lights and only when the RF protocol only has a "ToggleLight" signal available. So, unrelated to shades.]

**Comment by gymrat801** [comment]
I would say the educated guess would be the plugin, because other plugins that work with my Somfy shades do not act like this. The previous Homebridge plugins I've used all send a Close or Open via HomeKit no matter what the assumed/stored state is, so that if there's a mismatch in the true state and stored state, you still get the desired result. 

Update:
This is definitely a bug, as the Somfy setup doesn't work correctly in any fashion, even if HomeKit is the only method in which you control your shades. Even using the "Preset" screws up HomeKit's ability to control the shade. Example:
- Shade is closed and Home app shows shade is closed.
- Using the Home app you activate the "Preset" switch, powering it "On", to lift the shade to the preset level.
- Home now shows the shade as open...which it is.
- Tell Siri or try to use the Home app to open the shade. Siri fails. There's no way in the app to do it as the shade is already open.
- The only way to get Siri/Home app to work now is to send the close command followed by the open command.

The same thing happens if you use a remote or wall control that interferes with the saved state of the shade. Again, this is not the case with the Homebridge-MyLink plugins.

**Comment by gymrat801** [comment]
Submitted Issue #200, with logs and examples from Homebridge-MyLink which functions more usefully.

**Comment by thewolfanator** [comment]
yes, I am using Somfy RTS shades with Bond and then with this plugin to homebridge.  We need something badly that will report what state the blind is in!

**Reply by gymrat801** [reply]
> yes, I am using Somfy RTS shades with Bond and then with this plugin to homebridge. We need something badly that will report what state the blind is in!

The main problem is the communication to the motor is not 2-way. The motors have no way to tell the shades what state they are in. Currently HomeKit is assuming a state, and if Bond or Somfy remotes etc are used instead to change that state then HomeKit is out of sync and assumes the incorrect state. 

If you close your shades at night with a HomeKit scene, and then open them in the morning with the Somfy remote, HomeKit still thinks they’re closed, and there’s no way to ask the motor for it’s true state. Then when you go to close the shade again with a HomeKit scene HomeKit already thinks the shade is closed when it’s not, and doesn’t process the close command. 

The Somfy plug-in gets around this by having a slider where you can select the percentage you want your shade open. For some reason HomeKit will always process a percentage request, even if the stored state is already the same as the request. The Bond plug-in would need to check out the way the Somfy MyLink plug-in is configured, because while you can only correctly select 0%, 100% and 50% (= preset “My” button), it at least forces the request thru HomeKit.

**Comment by thewolfanator** [comment]
Yes, this is exactly what I thought too.  If the regular remote is used the plugins have no idea if the blind is open or closed.   I think the Tahoma base stations work differently.  Let say its sunset and blinds are programmed to go down.  It doesn't check the state of the blind and just sends the command anyways -  blinds that are down  stay down and blinds that are up now go down.  I think the problem is if homekit thinks the blind is already down it doesn't initiate the command to lower.  I wonder if there is any way around this?

Maybe set up automation in homekit for sunset - first open the blinds and then close 1 second later?  This way if the blind is in a closed state it will move ever so slightly up and then get a down command or maybe they won't even have time to move up at all?  I sometimes just do this manually by sliding the blind up and then down right away.  Maybe opposite would work for raising blinds?

**Reply by gymrat801** [reply]
> Yes, this is exactly what I thought too. If the regular remote is used the plugins have no idea if the blind is open or closed. I think the Tahoma base stations work differently. Let say its sunset and blinds are programmed to go down. It doesn't check the state of the blind and just sends the command anyways - blinds that are down stay down and blinds that are up now go down. I think the problem is if homekit thinks the blind is already down it doesn't initiate the command to lower. I wonder if there is any way around this?
> 
> Maybe set up automation in homekit for sunset - first open the blinds and then close 1 second later? This way if the blind is in a closed state it will move ever so slightly up and then get a down command or maybe they won't even have time to move up at all? I sometimes just do this manually by sliding the blind up and then down right away. Maybe opposite would work for raising blinds?

The problem with the automation idea, besides being cumbersome, is that you can’t set an automation in Homekit that way. You can’t use a device twice and you can’t set a timer between things. You can’t even set the order of events. You can trigger a scene or trigger additional devices not in a scene, but you can’t do what you describe. 

I’d have to go into IFTTT and build it there. But that negates the whole purpose of Bond because IFTTT can communicate with Somfy MyLink.

**Comment by chrismerck** [comment]
FWIW, a few comments here:

1. Bond Bridge does not check the current state of the shade before sending a signal. So if the app or an integration (such as HomeBridge) asks Bond Bridge to "Close" the shade, we will send a Close signal, regardless of the state that the Bridge believes the shade to be in.
2. Bond Bridge does report state back to the integration (HomeBridge). Our API is two-way, but the underlying RF technology is 1-way. So, indeed if the Somfy remote is used, then our state report will be wrong.
3. We have a solution for (2)! A forthcoming product, recently announced, will allow you to replace your Somfy remotes will our Sidekick keypads which control the shade and also communicate to the Bridge to update the state to integrations. -- Sure, it's still possible that a signal is missed, but the RF layer is pretty reliable.

**Reply by chrismerck** [reply]
[Sorry some of this was redundant. I totally forgot that I already commented above 🤦 ]

**Reply by gymrat801** [reply]
> FWIW, a few comments here:
> 
> 1. Bond Bridge does not check the current state of the shade before sending a signal. So if the app or an integration (such as HomeBridge) asks Bond Bridge to "Close" the shade, we will send a Close signal, regardless of the state that the Bridge believes the shade to be in.
> 2. Bond Bridge does report state back to the integration (HomeBridge). Our API is two-way, but the underlying RF technology is 1-way. So, indeed if the Somfy remote is used, then our state report will be wrong.
> 3. We have a solution for (2)! A forthcoming product, recently announced, will allow you to replace your Somfy remotes will our Sidekick keypads which control the shade and also communicate to the Bridge to update the state to integrations. -- Sure, it's still possible that a signal is missed, but the RF layer is pretty reliable.

How does the new device integrate with Homekit to get around the issue with shade state and the preset button not affecting state correctly?

Seems like the new device would suffer from the same issue if Homekit is involved. 

I think the only real solution is to change Homebridge-Bond to a full 0-100% slider as described above. No secondary “stop/preset” toggle accessory. Just like I described Homebridge-MyLink does. 

0-49% = Open
50% = preset
51-99% = Open
100% = Closed

**Comment by obwielnls** [comment]
I just ran into this same issue.   It would be great if it would just send a close or open command regardless of what current state it thinks the shades are in.   That would really solve the entire issue for me.

**Reply by OmriMulti** [reply]
This is the solution!!! Can someone make it happen?
